### ISSUE TITLE
Create a Task in Managed Service

### ISSUE THREAD

**Issue by ChrisKellaleaMaynard** [aehrc/snomio#343]
**As a:** Terminology Author,  
**I want to:**  create a Task in Managed Service via Snomio
**So I can:**  reduce the need for application-switching from my BAU.

---
### Acceptance Criteria
1. When the user views the "All Tasks" or "My Tasks" screens, a _Create Task_ button is displayed at the top of the list.
2. When the user clicks the _Create Task_ button, a modal window appears requesting the following base Task information from the user:
* Task Title  (single line of text)
* Project (drop-down list, read-only field, defaulted to "AUAMT")
* Number of tasks (single line of text, read-only field, defaulted to "1")
* Task Details  (multi-line text)
4. After entering the base Task information the user can click a _Create Task_ button and a Task is created within Managed Service for the AUAMT project, and pre-populated with the base Task information gathered.
5. After creating the Task, the base Task information window is hidden and the user is taken to the Task editing screen in Snomio with the new Task's details displayed (as if the user had just selected the Task from the "All Tasks" screen).
6. When the modal window requesting the base Task information is displayed, the user can click a _Cancel_ button to close the window without having a Task created.
7. If the user clicks the _Create Task_ button in the base Task information window and they haven't entered data for one or more mandatory fields, an error message "Please enter all mandatory fields." is displayed and the user remains in the base Task information window.

### Testing
1. [ test case ]

**Comment by HymaKanchiAU** [comment]
@ChrisKellaleaMaynard, please let me know if the status should be In-progress or New as there is an issue that if the status is new branch cannot be created.

**Comment by clinnygee** [comment]
To make it more clear, the issue is not that tasks are created as new - this is how they are created in the AP. If you create a task in the AP without opening it you encounter the same issue. 

These tasks are created identically to ones that are created in the authoring platform. 

Making these tasks created as in progress with a branch does not solve the issue, it just stops it happening for tasks that are created in snomio. The same issue can/will be encountered if you create tasks in the authoring platform.

The fix for the issue is to check if a branch exists when a task is opened in snomio, and if not, not create it.

**Comment by ChrisKellaleaMaynard** [comment]
The vast majority of tasks created via Snomio will want to be "In Progress", as the authors will spend time editing and reviewing the ticket for a while before deciding to move forward with the actual work - at which point they'll create the task and start entering product details.  This will be true even if they decide to use the "do everything in the atomic data entry form instead of putting details in the ticket first" approach, as even that approach means they'll create the task immediately before entering data.

The Content Team is aware that the earlier they create the task, the higher the potential for rebasing issues later in the process.  For our use cases, there is no reason to believe a user would ever want to create a task and not have the branch exist for it.  The branchless task may work for NRCs who don't have a ticketing system on top of the Managed Service, as the name-only tasks could be used for work management.  However we have a different workflow, thanks to the TGA/PBS data feeds.

Therefore when the task is created it should immediately have an "In progress" status, and a branch should be created.  When the user opens a task screen in Snomio it should also check for the existence of a branch and create it if necessary.  It is unlikely a user will create a task in AP and then add to it from Snomio, however adding that check now will prevent issues from potential future workflows.  Adding the check **does not** mean the branch creation can be skipped during task creation (in Snomio), as that solution is conceptually challenging compared to just telling users "when you create a task, the branch is also created".

**Comment by HymaKanchiAU** [comment]
Create Task option displayed on
My Tasks = Pass
All Tasks = Pass
Additionally present on Task requiring review = Fail - 
![image](https://github.com/aehrc/snomio/assets/126848487/1c5c2b28-e50e-4858-8c39-93481e404f88)

Verify the fields displayed on the create task modal. = Pass
![image](https://github.com/aehrc/snomio/assets/126848487/7379bfab-001e-4582-aeb4-e8ba7852f460)
Click Cancel button on the modal - should not create any tasks and the modal should be closed. = Pass
Click Esc on the modal - should not create any tasks and the modal should be closed. = Pass
Error message displayed when title is empty.=Pass
![image](https://github.com/aehrc/snomio/assets/126848487/48528545-5617-4eeb-a239-da9614215adc)

Create a task by entering a title with special characters with the default values - status should be In Progress and should display the task edit screen. = Partial Pass
![image](https://github.com/aehrc/snomio/assets/126848487/66753165-e6b4-406f-963e-8ed02f109320)

Receiving console errors 
![image](https://github.com/aehrc/snomio/assets/126848487/9a454f16-2954-4723-bf45-f283700ce425)
![image](https://github.com/aehrc/snomio/assets/126848487/875ee59b-5923-49ca-98ae-5dd0c57431d9)


Create multiple tasks - Status of the 1st task should be In-Progress and the remaining tasks should have status as New. 1st task should be similar to the task edit screen view. = Fail

Creating the specified number of tickets with the status as new.

Steps 2 to 7 are performed from both My Tasks and All tasks page. = Pass

In addition, 
Verified that the task status is showing as in-progress in AP = Pass
![image](https://github.com/aehrc/snomio/assets/126848487/2439e846-43ff-43ca-a51d-598696f7e135)

Verified that the details entered in the create task screen are shown in AP task details screen =Pass
![image](https://github.com/aehrc/snomio/assets/126848487/36abc072-b0af-494e-a2cc-f0a1b027473d)

Verified that able to search for a concept in the create product screen = Pass
![image](https://github.com/aehrc/snomio/assets/126848487/5ba33b20-9329-4b06-b4ad-fc09e6818e79)

![image](https://github.com/aehrc/snomio/assets/126848487/59c1c86f-fa01-449b-bbd7-df0f6a2b0b35)

![image](https://github.com/aehrc/snomio/assets/126848487/be317674-35bf-492f-929e-21a445c8e278)

**Comment by senjo-aehrc** [comment]
@HymaKanchiAU , Multi task creation is not a requirement in this ticket, Not sure how did you get this test scenario **"Create multiple tasks - Status of the 1st task should be In-Progress and the remaining tasks should have status as New. 1st task should be similar to the task edit screen view."** .  @ChrisKellaleaMaynard , How do you want to handle multi task creation?.
Also create button appears on the task requiring review page I think its for consistency we have to keep that button in all the tasks page. @ChrisKellaleaMaynard please advise if we need to hide it from that page.

**Comment by ChrisKellaleaMaynard** [comment]
These are both good behaviours to capture as part of testing!

The _Create Task_ button should be visible on each of the pages where there is a list of tasks.  This assists with reinforcement to the user that they're still looking at tasks, and haven't switched to tickets.  When additional task-management functions are added in the future, they could all be placed on a tollbar that appears whenever the user is viewing lists of tasks - like the table editing tools that appear and disappear in Confluence.

With the multi-task creation, each of the new tasks should be created in the same manner as if a single task was being created: **Status** = "In Progress", branch is created, etc.  When a single task is created the user is immediately transferred to the task editing screen, therefore the same should happen when creating multiple tasks simultaneously: the first task should be shown.  In the vast majority of cases, even if the user has the intention to modify each task (e.g. by adding tickets to them) the first task created is the most likely starting point.

**Comment by senjo-aehrc** [comment]
@ChrisKellaleaMaynard , I have noticed a behaviour with multi task creation some times the task ID is not in the right order with the task name. This is because of the parallel async api call to the authoring  service to create multiple tasks.
![image](https://github.com/aehrc/snomio/assets/135300932/28e4a03d-3d51-4928-bdaf-5ec831ad4b07)
Are we really need to maintain the order of the id with the name, If required we need to change this to the sequential

**Comment by ChrisKellaleaMaynard** [comment]
No, that's not a priority: in the odd occasion where someone decides they really need to create multiple tasks at once, they can always change the title from "#1" to "#3" if necessary - but it's way more likely they would change the title from "#1" to "3-pack", "#2" to "6-pack", and "#3" to "12-pack".  The majority of tasks will be created one at a time, so the documentation will just need to reflect this potential for a different task numbering order.  Ultimately the task numbers are only unique identifiers, not indicators of the order in which things are built.  I'll have to remember that line for the documentation.

**Comment by senjo-aehrc** [comment]
Awesome, Thanks @ChrisKellaleaMaynard

**Comment by senjo-aehrc** [comment]
@HymaKanchiAU , Its ready for testing again. Please use the latest build

**Comment by senjo-aehrc** [comment]
@HymaKanchiAU , Noticed one issue, Please hold on for this testing
